### PR TITLE
[ArrayAndCollectionExtensions]

### PR DESCRIPTION
**Pull request by dyrkabes**
Extended Collection protocol with [safe:] subscription that either returns the element with the index or nil if the index does not exist
Added distinct() extension to Array which removes all the duplicate elements saving the order of elements

P.S. That's my first open source commit, looking forward to any comments :)

### CODE REVIEW

File: HandySwift.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -7,6 +7,10 @@
 	objects = {
 
 /* Begin PBXBuildFile section */
+		3F95C8D220F22A3C0045AFD0 /* ColletionExtension.swift in Sources */ = {isa = PBXBuildFile; fileRef = 3F95C8D120F22A3C0045AFD0 /* ColletionExtension.swift */; };
```
Comment (Dschee) [inline HandySwift.xcodeproj/project.pbxproj]: You mistyped `Collection` by forgetting the second `c`. 🙂 
Please fix.

File: README.md [hunk 2]
```diff
@@ -97,6 +97,7 @@ Open the Playground from within the `.xcworkspace` in order for it to work.
   - [ArrayExtension](#arrayextension)
   - [DictionaryExtension](#dictionaryextension)
   - [DispatchTimeIntervalExtension](#dispatchtimeintervalextension)
+  - [CollectionExtension](#collectionextension)
```
Comment (Dschee) [inline README.md]: I think, from the order perspective, it makes more sense that `Collection` appears somewhere around array. I think I'd place it before `Array` even.

File: README.md [hunk 3]
```diff
@@ -212,6 +213,14 @@ Combines each element with each element of a given other array.
 // => [(1, "A"), (1, "B"), (2, "A"), (2, "B"), (3, "A"), (3, "B")]
 ```
 
+#### .distinct()
```
Comment (Dschee) [inline README.md]: I'm not sure about the naming, I'm used to something like `uniq()` from Ruby. But `distinct()` would be okay too. The bigger concern from my point of view is: Why not use a `Set` istead if I want uniq objects? I mean, that's what it was made for. And if you need an ordered set, I'd rather introduce a `OrderedSet` type instead of adding a `.distinct()` method to Swift. But that could be done in a different PR.

File: Sources/Extensions/ArrayExtension.swift [hunk 4]
```diff
@@ -109,6 +109,21 @@ extension Array {
     }
 }
 
+extension Array where Element : Equatable {
```
Comment (Dschee) [inline Sources/Extensions/ArrayExtension.swift]: Please remove this from this PR for now. See my comment above.

File: Sources/Extensions/ColletionExtension.swift [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+//  HandySwift iOS
+//
+//  Created by Stepanov Pavel on 08/07/2018.
+//  Copyright © 2018 Flinesoft. All rights reserved.
+//
+
+import Foundation
+
+extension Collection {
+
```
Comment (Dschee) [inline Sources/Extensions/ColletionExtension.swift]: This empty line should actually have shown up as a warning if you have SwiftLint installed on your Mac. Please remove it.

File: Sources/Extensions/ColletionExtension.swift [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+//
+//  Created by Stepanov Pavel on 08/07/2018.
+//  Copyright © 2018 Flinesoft. All rights reserved.
+//
+
+import Foundation
+
+extension Collection {
+    
+    /// Returns an element with the specified index or nil if the element does not exist
```
Comment (Dschee) [inline Sources/Extensions/ColletionExtension.swift]: Please set a `.` at the end of this line since this is a valid english sentence.

File: Sources/Extensions/ColletionExtension.swift [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+    ///
+    /// - Parameters:
+    ///   - safe: The index of the element
```
Comment (Dschee) [inline Sources/Extensions/ColletionExtension.swift]: Same here, please add punctuation at end of line.

File: Sources/Extensions/ColletionExtension.swift [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    public subscript (safe index: Index) -> Element? {
```
Comment (Dschee) [inline Sources/Extensions/ColletionExtension.swift]: - [ ] Please also check this line with SwiftLint installed (the whitespace between `subscript` and `(safe` should show a warning).
- [ ] We already wanted to add such a subscript to this project with some colleges and discussed, how this subscript should be named. We agreed on the subscript name `(try` instead of `(safe`. If you're okay with that name, too, could you rename your implementation?

File: Tests/Extensions/ArrayExtensionTests.swift [hunk 9]
```diff
@@ -73,4 +73,11 @@ class ArrayExtensionTests: XCTestCase {
             XCTAssertEqual(testArray[index], sortedArray[index])
         }
     }
+    
+    func testDistinct() {
+        let testArray = [0, 1, 2, 2, 2, 3, 4, 5, 6, 6]
+        let resultArray = [0, 1, 2, 3, 4, 5, 6]
+
```
Comment (Dschee) [inline Tests/Extensions/ArrayExtensionTests.swift]: Here, tests for edge cases like empty arrays would be important. But according to my above comments, this should be removed from this PR anyways – but for the future...

**Comment by dyrkabes** [comment]
Thank you for such a comprehensive and fast review, I will fix the issues and push a bit later. I couldn't even think about some of the comments so it's a great experience.

And I will install SwiftLint.. for some reasons haven't used it :)

**Comment by dyrkabes** [comment]
By the way, there's an issue with the playground file, it does not compile at the regex part. It says "Use of unresolved identifier 'regex'". I suppose that it is due to some renaming... probably there should be a  separate PR with fix
